Add `process_density_matrix` for classical shadow

JerryChen97 opened the pull request:
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
Classical Shadow used to have two methods to treat statevectors and density matrices in completely different api: `process_state_with_shots` is the clean version compatible with new device API and equipped with faster algorithm using iterative methods, but specifically only works for statevectors; 'process' is the general version which also works for density matrices but has much slower algorithm and meanwhile actually only legacy devices are calling this interface.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-81306]

astralcai commented on pennylane/measurements/classical_shadow.py:
```diff
@@ -442,6 +442,131 @@ def process_state_with_shots(
+
+        # seed the random measurement generation so that recipes
+        # are the same for different executions with the same seed
+        recipe_rng = np.random.RandomState(self.seed)
```
Why do we use `self.seed` here but `rng` from arguments a few lines below?

dwierichs replied:
I think it makes sense because there are two sources of randomness: what to measure (purely classical) and the measured outcome (quantum on QC, classical in simulators)

dwierichs commented on pennylane/measurements/classical_shadow.py:
```diff
@@ -442,6 +442,131 @@ def process_state_with_shots(
+
+        bit_rng = np.random.default_rng(rng)
+
+        obs_list = np.stack(
+            [
+                qml.X.compute_matrix(),
+                qml.Y.compute_matrix(),
+                qml.Z.compute_matrix(),
+            ]
+        )
+
+        # the diagonalizing matrices corresponding to the Pauli observables above
+        diag_list = np.stack(
+            [
+                qml.Hadamard.compute_matrix(),
+                qml.Hadamard.compute_matrix() @ qml.RZ.compute_matrix(-np.pi / 2),
+                qml.Identity.compute_matrix(),
+            ]
+        )
```
I'd almost be drawn to suggest that we hard-code these matrices :sweat_smile: 
```suggestion
        obs_list = np.array([
            [[0, 1], [1, 0]],
            [[0, -1j], [1j, 0]],
            [[1, 0], [0, -1]],
        ])

        # the diagonalizing matrices corresponding to the Pauli observables above
        diag_list = np.array([
            [[1 / np.sqrt(2), 1 / np.sqrt(2)], [1 / np.sqrt(2), -1 / np.sqrt(2)]],
            [[0.5 + 0.5j, 0.5-0.5j], [0.5 + 0.5j, -0.5+0.5j]],
            [[1, 0], [0, 1]],
        ])
```

dwierichs commented on pennylane/measurements/classical_shadow.py:
```diff
@@ -442,6 +442,131 @@ def process_state_with_shots(
+        )
+
+        outcomes = np.zeros((shots, n_qubits))
+        stacked_state = np.repeat(transposed_state[np.newaxis, ...], shots, axis=0)
```
Wait, are we producing a new copy of the state for every shot here? Is that really necessary? :sweat_smile:

dwierichs replied:
Ah yeah I suppose it is... How is this comparing memory footprint wise to the previously used method? :)

JerryChen97 replied:
Good idea! I'll compare the ram usage soon as well

JerryChen97 replied:
Marked as done. Comparison in the figure posted as a new conversation.

dwierichs commented on pennylane/measurements/classical_shadow.py:
```diff
@@ -442,6 +442,131 @@ def process_state_with_shots(
+            rho_str = f"{stacked_dim}a{traced_dim}{a_dagger_dim}..."
+            UT_str = f"{stacked_dim}{a_dagger_dim}{v_dim_right}"
+            new_str = f"{stacked_dim}{v_dim_left}{traced_dim}{v_dim_right}..."
+            rotated_state = np.einsum(
```
Could it be worth to figure out the optimal contraction order for this one and hardcode it in?

JerryChen97 replied:
That's a good idea! Let me find some time to figure this out

dwierichs replied:
Might be fine not doing that here and now :)

JerryChen97 replied:
Yeah actually I just quickly went through `einsum`'s own `greedy` and `optimal`, both are not introducing very obvious improvement. Probably better we backlog this for future investigation.

JerryChen97 replied:
Their `optimal` is even having performance downgrade which I guess is caused by their own theoretical solver is using comparable time as the whole shadow algorithm.

JerryChen97 replied:
[sc-91553]

dwierichs commented on tests/measurements/test_classical_shadow.py:
```diff
@@ -223,6 +223,159 @@ def test_expval_same_rng(self):
 
 
+class TestProcessDensityMatrix:
+    """Unit tests for process_state_with_shots for the classical_shadow
```
```suggestion
    """Unit tests for process_density_matrix_with_shots for the classical_shadow
```

dwierichs commented on tests/measurements/test_classical_shadow.py:
```diff
@@ -223,6 +223,159 @@ def test_expval_same_rng(self):
+        assert res.dtype == np.int8
+
+        # test that the bits are either 0 and 1
+        assert np.all(np.logical_or(res[0] == 0, res[0] == 1))
```
Does this work?
```suggestion
        assert set(res[0]).issubset({0, 1})
```

dwierichs commented on tests/measurements/test_classical_shadow.py:
```diff
@@ -223,6 +223,159 @@ def test_expval_same_rng(self):
+        assert np.all(np.logical_or(res[0] == 0, res[0] == 1))
+
+        # test that the recipes are either 0, 1, or 2 (X, Y, or Z)
+        assert np.all(np.logical_or(np.logical_or(res[1] == 0, res[1] == 1), res[1] == 2))
```
```suggestion
        assert set(res[1]).issubset({0, 1, 2})
```

dwierichs commented on tests/measurements/test_classical_shadow.py:
```diff
@@ -223,6 +223,159 @@ def test_expval_same_rng(self):
+        # test that the first qubit samples are all 0s when the recipe is Z
+        assert np.all(res[0][res[1, ..., 0] == 2][:, 0] == 0)
+
+        # test that the second qubit samples contain 1s when the recipe is Z
+        assert np.any(res[0][res[1, ..., 1] == 2][:, 1] == 1)
```
Should we fix seeds if we rely on this random sampling yielding a 1?

JerryChen97 replied:
seeded at https://github.com/PennyLaneAI/pennylane/pull/6748/commits/29f32747b92727108f267fb47a39f12b3186ebdd and https://github.com/PennyLaneAI/pennylane/pull/6748/commits/6c7a834d6dea7675507e2d825a18e7b01ca184d9

dwierichs replied:
Thanks, amazing :)

dwierichs commented on tests/measurements/test_classical_shadow.py:
```diff
@@ -223,6 +223,159 @@ def test_expval_same_rng(self):
+        assert res.dtype == np.int8
+
+        # test that the bits are either 0 and 1
+        assert np.all(np.logical_or(res[0] == 0, res[0] == 1))
+
+        # test that the recipes are either 0, 1, or 2 (X, Y, or Z)
+        assert np.all(np.logical_or(np.logical_or(res[1] == 0, res[1] == 1), res[1] == 2))
```
```suggestion
        assert set(res[0]).issubset({0, 1})

        # test that the recipes are either 0, 1, or 2 (X, Y, or Z)
        assert set(res[1]).issubset({0, 1, 2})
```

dwierichs reviewed (changes requested):
It could be good to generalize some test to more than two wires, if possible (maybe with very small shot count to save cost) and confirm wire ordering still works. Two can be a bit of a special number when checking against accidental permutations and such bugs :D

JerryChen97 commented:
I'm thinking of either just drop the test for `test_hadamard_expval_mixed` or completely redirect `qml.shadow_expval` to the new method for `default.mixed`. Right now it looks like to me that method is still using the old method.

JerryChen97 commented:
Yeah right now it's much faster (ignored the wrong branch names they were caused by my lagged update):
```bash
(venv) ➜  pennylane git:(classical-shadow/add-process-densitymatrix) ✗ pytest tests/measurements/test_classical_shadow.py::test_hadamard_expval_mixed --durations=5
==================================== test session starts =====================================
platform linux -- Python 3.11.11, pytest-8.3.5, pluggy-1.5.0
benchmark: 5.1.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /home/yushao.chen/Documents/pennylane/tests
configfile: pytest.ini
plugins: forked-1.6.0, split-0.10.0, nbmake-1.5.5, cov-6.1.1, mock-3.14.0, rng-1.0.0, anyio-4.8.0, flaky-3.8.1, xdist-3.6.1, timeout-2.4.0, benchmark-5.1.0
collected 1 item                                                                             

tests/measurements/test_classical_shadow.py .                                          [100%]

==================================== slowest 5 durations =====================================
87.56s call     measurements/test_classical_shadow.py::test_hadamard_expval_mixed

(2 durations < 0.005s hidden.  Use -vv to show these durations.)
================================ 1 passed in 87.92s (0:01:27) ================================
(venv) ➜  pennylane git:(master) ✗ pytest tests/measurements/test_classical_shadow.py::test_hadamard_expval_mixed --durations=5
=============================================================== test session starts ===============================================================
platform linux -- Python 3.11.11, pytest-8.3.5, pluggy-1.5.0
benchmark: 5.1.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /home/yushao.chen/Documents/pennylane/tests
configfile: pytest.ini
plugins: forked-1.6.0, split-0.10.0, nbmake-1.5.5, cov-6.1.1, mock-3.14.0, rng-1.0.0, anyio-4.8.0, flaky-3.8.1, xdist-3.6.1, timeout-2.4.0, benchmark-5.1.0
collected 1 item                                                                                                                                  

tests/measurements/test_classical_shadow.py .                                                                                               [100%]

=============================================================== slowest 5 durations ===============================================================
2.78s call     measurements/test_classical_shadow.py::test_hadamard_expval_mixed

(2 durations < 0.005s hidden.  Use -vv to show these durations.)
================================================================ 1 passed in 3.11s ================================================================
(venv) ➜  pennylane git:(classical-shadow/add-process-densitymatrix) ✗ 
```

JerryChen97 commented on pennylane/measurements/classical_shadow.py:
```diff
@@ -442,6 +442,131 @@ def process_state_with_shots(
         return np.stack([outcomes, recipes]).astype(np.int8)
 
+    def process_density_matrix_with_shots(
+        self, state: Sequence[complex], wire_order: Wires, shots: int, rng=None
```
Just to align with method `process_state_with_shots` above lol

JerryChen97 commented:
Should be fine as long as below 2GB

astralcai commented:
> ![image](https://private-user-images.githubusercontent.com/21050723/445708064-c2d3ebed-dd4c-4e56-93b8-c4139d21955c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc3Njc1ODUsIm5iZiI6MTc0Nzc2NzI4NSwicGF0aCI6Ii8yMTA1MDcyMy80NDU3MDgwNjQtYzJkM2ViZWQtZGQ0Yy00ZTU2LTkzYjgtYzQxMzlkMjE5NTVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTIwVDE4NTQ0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFlYzU1MWM2NjExZWJkNWViMmQ1ZGZmNmNmODgzOTk4YTUwNTUyNDE1ZTkyNzJiZTQ2YTM0Mzc4NDQ1N2NhOTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Phlroa0ICCqXxzqyVMVYMZuDw0LduOrt0S7IpD-kGn4)
> 
> Mem used increased!!! @dwierichs @astralcai

I think this looks acceptable.

astralcai commented on pennylane/measurements/classical_shadow.py:
```diff
@@ -442,6 +442,131 @@ def process_state_with_shots(
+        bit_rng = np.random.default_rng(rng)
+
+        obs_list = np.array(
+            [
+                [[0, 1], [1, 0]],
+                [[0, -1j], [1j, 0]],
+                [[1, 0], [0, -1]],
+            ]
```
Can you add comments for what each of them is?

JerryChen97 replied:
https://github.com/PennyLaneAI/pennylane/pull/6748/commits/1efa0472312e6711c2924c745ba391dec8aa64aa

JerryChen97 commented on pennylane/devices/qubit_mixed/sampling.py:
```diff
@@ -154,95 +154,32 @@ def _measure_classical_shadow(
-            rho_q = math.reduce_dm(rho_matrix, [q_wire])
-
-            # (B) rotate that 2x2 block to Z-basis if recipe is X or Y
-            recipe = recipes[t, q_idx]
-            U = diag_list[recipe]
-            U = math.convert_like(U, rho_q)
-            U_dag = math.conjugate(math.transpose(U))
-            rotated = math.dot(U, math.dot(rho_q, U_dag))
-
-            # (C) probability of outcome 0 => rotated[0,0].real
-            p0 = math.clip(math.real(rotated[0, 0]), 0.0, 1.0)
-            if bit_rng.random() < p0:
-                outcomes[t, q_idx] = 0
-            else:
-                outcomes[t, q_idx] = 1
-
-    res = math.stack([outcomes, recipes]).astype(np.int8)
-    return res
+    return mp.process_density_matrix_with_shots(
+        state,
+        wire_order,
+        shots,
+        rng=rng,
+    )
```
@astralcai @dwierichs Note that I also modified the workflow of `default.mixed` s.t. it directly uses our new routine now. Or some tests along with the `qml.shadow_expval` will still use code here same as the legacy code to be super slow.

dwierichs replied:
Sounds good! What exactly was going on with the `is_state_batched` argument before? :sweat_smile: Was it just always `False`? :thinking:

JerryChen97 commented on tests/measurements/test_classical_shadow.py:
```diff
@@ -223,6 +224,197 @@ def test_expval_same_rng(self):
+        # now test that the second qubit samples are all 0s when the recipe is Z
+        assert np.all(res[0][res[1, ..., 1] == 2][:, 1] == 0)
+
+    @pytest.mark.parametrize("num_wires", [3, 4, 5])
+    @pytest.mark.parametrize("shots", [100])
+    def test_wire_order_generalized(self, num_wires, shots, seed):
+        """Test that the wire order is respected for any number of wires and all permutations"""
+        # Prepare a GHZ-like state for num_wires
+        state = np.zeros((2,) * num_wires)
+        idx0 = tuple([0] * num_wires)
+        idx1 = tuple([1] * num_wires)
+        state[idx0] = 1 / np.sqrt(2)
+        state[idx1] = 1 / np.sqrt(2)
+        dm = np.outer(state.ravel(), state.ravel()).reshape((2,) * (2 * num_wires))
+
+        wires = list(range(num_wires))
+        mp = qml.classical_shadow(wires=wires)
+
+        for perm in itertools.permutations(wires):
+            res = mp.process_density_matrix_with_shots(
+                dm, qml.wires.Wires(list(perm)), shots=shots, rng=seed
+            )
+            assert res.shape == (2, shots, num_wires)
+            # bits are always 0 or 1
+            assert set(res[0].ravel()).issubset({0, 1})
+            # recipes are always 0, 1, or 2
+            assert set(res[1].ravel()).issubset({0, 1, 2})
+
+            # For GHZ state, when all qubits are measured in Z basis,
+            # they should all have the same value (all 0s or all 1s)
+            z_shots = np.where(np.all(res[1] == 2, axis=1))[0]
+            if len(z_shots) > 0:  # Only test if we have any all-Z shots
+                for shot_idx in z_shots:
+                    # All qubits should have the same measurement outcome
+                    first_bit = res[0][shot_idx, 0]
+                    assert np.all(res[0][shot_idx] == first_bit)
```
@dwierichs I added a test here to make use of GHZ s.t. we can have deterministic results. However as I increased the `num_wires` the time cost still increased fast so stopped at `5`.

dwierichs replied:
Oh wow, yeah that's quite a general test :D I'm not surprised that this blows up very quickly: The density matrix scales as $4^n$, the number of contractions scales with $n$ and the number of permutations is $n!$, so that's a scaling of at least $n 4^n (n!)$, which is about as horrible as something can scale in real-world code, I suppose :D

dwierichs commented on pennylane/measurements/classical_shadow.py:
```diff
@@ -442,6 +442,131 @@ def process_state_with_shots(
+            )
+
+            # sample the observables on the first qubit
+            probs = (np.einsum("abc,acb->a", first_qubit_state, obs[:, active_qubit]) + 1) / 2
```
I was thinking: Currently we compute the probs in three different bases, then rotate the state, correct?
We could also just rotate first and the compute the probs in `Z` basis exclusively, reducing the second object in this contraction here from `(num_shots, 2, 2)` to `(2, 2)` :sweat_smile: 
But maybe that's an optimization for later :)

JerryChen97 replied:
Sounds a good idea! Will record this to the story for later optimization as well

dwierichs reviewed (approved):
Nice, thanks for the thorough review reply and generalized wire order test! :) 
Nice job @JerryChen97 :rocket:

astralcai reviewed (approved):
Thank you for doing this 🚀

JerryChen97 commented:
@dwierichs Previous code living inside `devices/qubit_mixed/sampling.py` was copied paste from the legacy code, and `is_state_batched` is something not needed there. It is here basically for interface consistency.
Indeed I directly used `permutations` so the final complexity exploded 😓 luckily `num_wires=5` didn't cause much trouble yet.

JerryChen97 commented:
@astralcai @dwierichs I'm thinking to cherry-pick this PR to a new one since I couldn't find any way to resolve the GitHub bug :sweat:

JerryChen97 commented:
in favor of https://github.com/PennyLaneAI/pennylane/pull/7458
